packages/react-test-renderer/src/ReactShallowRenderer.js

```diff
@@ -198,8 +203,11 @@ class ReactShallowRenderer {
             }
           }
         }
-        this._instance.componentWillMount();
-      } else {
+
+        if (typeof element.type.getDerivedStateFromProps !== 'function') {
+          this._instance.componentWillMount();
+        }
+      } else if (typeof element.type.getDerivedStateFromProps !== 'function') {
         this._instance.UNSAFE_componentWillMount();
       }
 
```

Comment:
I haven't looked through those changes in detail.

I think it would be good to write up a table of what happens when both of either of methods is present (e.g. gDSFP, cWM/u_cWM, cWRP/u_CRWP).

My concern is that maybe some edge cases (e.g. "both cWRP and gDSFP" or "both u_CWRP and gDSFP" etc) may be handled differently between Fiber, SSR, and Shallow renderer. Of course we'd warn in this case but consistent behavior is still important.

Note I'm not saying that there *are* mistakes. Just that this logic is getting complex, and it's not obvious to me that all these cases work the same way. Making a table with different permutations and expected result would help us fix bugs if we find any later. It might also be helpful for the blog post.

Reply:
> I think it would be good to write up a table of what happens when both of either of methods is present (e.g. gDSFP, cWM/u_cWM, cWRP/u_CRWP).

I have tests that codify that (`ReactComponentLifeCycle-test`, `ReactDOMServerLifecycles-test`, and `ReactShallowRenderer-test`) for all three renderers.

tl;dr is that none of the unsafe lifecycles are invoked if gDSFP is present. This is necessary for the polyfill approach to work.

Reply:
> Making a table with different permutations and expected result would help us fix bugs if we find any later. It might also be helpful for the blog post.

Where would such a table live? Seems like codified in tests is as good as any.